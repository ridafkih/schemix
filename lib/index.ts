import { PrismaSchema } from "modules/PrismaSchema";

import { PrismaDataSourceOptions } from "typings/prisma-datasource";
import {
  PrismaGeneratorOptions,
  PrismaMultiGeneratorOptions,
} from "typings/prisma-generator";

type PrivateSchema = Omit<
  PrismaSchema,
  "createMixin" | "createEnum" | "createModel"
>;

interface CreateSchemaOptions {
  datasource: PrismaDataSourceOptions;
  generator: PrismaGeneratorOptions | PrismaMultiGeneratorOptions;
  basePath?: string;
  additionalPaths?: string[];
}

export let schema: PrismaSchema | undefined;

/**
 * Create a Prisma schema object.
 * @param props.datasource The datasource object information for Prisma.
 * @param props.generator The generator object information for Prisma.
 * @param props.basePath The base path for the schema object, optional if you want a purely programmatically generated schema.
 * @param props.additionalPaths The path for a secondary location for the schema objects. Will not be searched if basePath is empty.
 * @returns The generated schema object.
 */
export const createSchema = <T extends CreateSchemaOptions>({
  datasource,
  generator,
  basePath,
  additionalPaths,
}: T): T["basePath"] extends string ? PrivateSchema : PrismaSchema => {
  if (typeof basePath === "undefined") {
    return new PrismaSchema(datasource, generator);
  }

  if (Array.isArray(additionalPaths)) {
    schema = new PrismaSchema(datasource, generator, basePath, additionalPaths);
  } else {
    schema = new PrismaSchema(datasource, generator, basePath);
  }

  return schema;
};

export { createMixin, createEnum, createModel } from "util/create";
export { extendModel } from "util/extend";
export { PrismaModel } from "modules/PrismaModel";
export { PrismaEnum } from "modules/PrismaEnum";
export { PrismaSchema };
