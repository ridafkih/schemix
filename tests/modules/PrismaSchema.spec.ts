import { PrismaSchema } from "schemix/lib/index";
import { describe, expect, it } from "vitest";

describe("PrismaSchema", () => {
  it("Should accept multiple generators", async () => {
    const schema = new PrismaSchema(
      {
        provider: "cockroachdb",
        url: { env: "DATABASE_URL" },
      },
      [
        {
          provider: "prisma-client-js",
          name: "client",
        },
        {
          provider: "other-client",
          name: "other",
        },
      ]
    );

    const asString = await schema.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should accept extensions", async () => {
    const schema = new PrismaSchema(
      {
        provider: "postgresql",
        url: { env: "DATABASE_URL" },
        extensions: ["pg_tgrm", "zombodb"],
      },
      {
        provider: "prisma-client-js",
      }
    );

    const asString = await schema.toString();
    expect(asString).toMatchSnapshot();
  });
  it("Should support shadowDatabaseUrl with environment variables", async () => {
    const schema = new PrismaSchema(
      {
        provider: "postgresql",
        url: { env: "DATABASE_URL" },
        shadowDatabaseUrl: { env: "SHADOW_DATABASE_URL" },
      },
      {
        provider: "prisma-client-js",
      }
    );

    const asString = await schema.toString();
    expect(asString).toMatchSnapshot();
  });
});
