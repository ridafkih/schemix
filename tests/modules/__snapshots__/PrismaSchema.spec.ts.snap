// Vitest Snapshot v1

exports[`PrismaSchema > Should accept extensions 1`] = `
"datasource database {
  provider   = \\"postgresql\\"
  url        = env(\\"DATABASE_URL\\")
  extensions = [pg_tgrm, zombodb]
}

generator client {
  provider = \\"prisma-client-js\\"
}
"
`;

exports[`PrismaSchema > Should accept multiple generators 1`] = `
"datasource database {
  provider = \\"cockroachdb\\"
  url      = env(\\"DATABASE_URL\\")
}

generator client {
  provider = \\"prisma-client-js\\"
}

generator other {
  provider = \\"other-client\\"
}
"
`;

exports[`PrismaSchema > Should support shadowDatabaseUrl with environment variables 1`] = `
"datasource database {
  provider          = \\"postgresql\\"
  url               = env(\\"DATABASE_URL\\")
  shadowDatabaseUrl = env(\\"SHADOW_DATABASE_URL\\")
}

generator client {
  provider = \\"prisma-client-js\\"
}
"
`;
